src/wp-includes/media.php

```diff
@@ -5468,53 +5550,91 @@ function wp_get_webp_info( $filename ) {
- * This default threshold of 3 content elements to omit the `loading` attribute for can be customized using the
- * {@see 'wp_omit_loading_attr_threshold'} filter.
- *
- * @since 5.9.0
+ * @since 6.3.0
  *
  * @global WP_Query $wp_query WordPress Query object.
  *
- * @param string $context Context for the element for which the `loading` attribute value is requested.
- * @return string|bool The default `loading` attribute value. Either 'lazy', 'eager', or a boolean `false`, to indicate
- *                     that the `loading` attribute should be skipped.
+ * @param string $tag_name The tag name.
+ * @param array  $attr     Array of the attributes for the tag.
+ * @param string $context  Context for the element for which the loading optimization attribute is requested.
+ * @return array Loading optimization attributes.
  */
-function wp_get_loading_attr_default( $context ) {
+function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
 	global $wp_query;
 
-	// Skip lazy-loading for the overall block template, as it is handled more granularly.
+	/*
+	 * Closure for postprocessing logic.
+	 * It is here to avoid duplicate logic in many places below, without having
+	 * to introduce a very specific private global function.
+	 */
+	$postprocess = function( $loading_attributes, $with_fetchpriority = false ) use ( $tag_name, $attr, $context ) {
+		if ( $with_fetchpriority ) {
+			$loading_attributes = wp_maybe_add_fetchpriority_high_attr( $loading_attributes, $tag_name, $attr );
+		}
+		return wp_maybe_strip_loading_attribute( $loading_attributes, $tag_name, $context );
+	};
+
+	$loading_attrs = array();
+
+	/*
+	 * Skip lazy-loading for the overall block template, as it is handled more granularly.
+	 * The skip is also applicable for `fetchpriority`.
+	 */
 	if ( 'template' === $context ) {
-		return false;
+		return $loading_attrs;
+	}
+
+	// For now this function only supports images and iframes.
+	if ( 'img' !== $tag_name && 'iframe' !== $tag_name ) {
+		return $loading_attrs;
+	}
+
+	// For any resources, width and height must be provided, to avoid layout shifts.
+	if ( ! isset( $attr['width'] ) || ! isset( $attr['height'] ) ) {
+		return $loading_attrs;
+	}
+
+	if ( isset( $attr['loading'] ) ) {
+		if ( 'lazy' === $attr['loading'] ) {
+			$loading_attrs['loading'] = $attr['loading'];
+			if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
+				_doing_it_wrong(
+					__FUNCTION__,
+					__( 'An image should not be lazy-loaded and marked as high priority at the same time.' ),
+					'6.3.0'
+				);
+			}
+		}
+
+		return $postprocess( $loading_attrs, true );
+	}
+
+	// An image with `fetchpriority="high"` cannot be assigned `loading="lazy"` at the same time.
+	if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
+		return $postprocess( $loading_attrs, true );
 	}
```

Comment:
```suggestion
	// An image with `fetchpriority="high"` cannot be assigned `loading="lazy"` at the same time.
	if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
	         if ( isset( $attr['loading'] ) && 'lazy' === $attr['loading'] ) {
	               $loading_attrs['loading'] = $attr['loading'];
	               _doing_it_wrong(
				__FUNCTION__,
				__( 'An image should not be lazy-loaded and marked as high priority at the same time.' ),
				'6.3.0'
			);
	         }
		return $postprocess( $loading_attrs, true );
	}
```

This logic is a little cleaner.

Reply:
@spacedmonkey issue with this change would be, `return $postprocess( $loading_attrs, true );` would not be called if only the `$attr['loading']` is set. We will still want to return early if only the loading attribute is present and also we want to check if it can be set fetchpriority high.

Reply:
@spacedmonkey What @kt-12 says, the logic you're proposing is not the same as what's currently there, in your proposal the early return in case of `isset( $attr['loading'] )` is missing.